cmd/serverless-init/main.go

```diff
@@ -8,108 +8,126 @@
-)
+const datadogConfigPath = "datadog.yaml"
 
-type cliParams struct {
-	args []string
-}
+var modeConf mode.Conf
 
 func main() {
-	if len(os.Args) < 2 {
-		panic("[datadog init process] invalid argument count, did you forget to set CMD ?")
-	}
 
-	cliParams := &cliParams{
-		args: os.Args[1:],
-	}
-
-	err := fxutil.OneShot(run, fx.Supply(cliParams))
+	modeConf = mode.DetectMode()
+	setEnvWithoutOverride(modeConf.EnvDefaults)
+
+	err := fxutil.OneShot(
+		run,
+		autodiscoveryimpl.Module(),
+		workloadmetafx.Module(),
+		fx.Provide(func(config coreconfig.Component) healthprobeDef.Options {
+			return healthprobeDef.Options{
+				Port:           config.GetInt("health_port"),
+				LogsGoroutines: config.GetBool("log_all_goroutines_when_unhealthy"),
+			}
+		}),
+		taggerimpl.Module(),
+		healthprobeFx.Module(),
+		fx.Supply(workloadmeta.NewParams()),
+		fx.Supply(tagger.NewTaggerParams()),
+		fx.Supply(core.BundleParams{
+			ConfigParams: coreconfig.NewParams("", coreconfig.WithConfigMissingOK(true)),
+			SecretParams: secrets.NewEnabledParams(),
+			LogParams:    logimpl.ForOneShot(modeConf.LoggerName, "off", true)}),
+		core.Bundle(),
+	)
 
 	if err != nil {
-		logger.Error(err)
+		log.Error(err)
+		os.Exit(-1)
 	}
 }
 
-func run(cliParams *cliParams) {
-	cloudService, logConfig, traceAgent, metricAgent, logsAgent := setup()
-	initcontainer.Run(cloudService, logConfig, metricAgent, traceAgent, logsAgent, cliParams.args)
-}
+// removing these unused dependencies will cause silent crash due to fx framework
+func run(_ secrets.Component, _ autodiscovery.Component, _ healthprobeDef.Component) {
+	cloudService, logConfig, traceAgent, metricAgent, logsAgent := setup(modeConf)
 
-func setup() (cloudservice.CloudService, *log.Config, *trace.ServerlessTraceAgent, *metrics.ServerlessMetricAgent, logsAgent.ServerlessLogsAgent) {
-	if err := config.SetupLogger(
-		loggerName,
-		"error", // will be re-set later with the value from the env var
-		"",      // logFile -> by setting this to an empty string, we don't write the logs to any file
-		"",      // syslog URI
-		false,   // syslog_rfc
-		true,    // log_to_console
-		false,   // log_format_json
-	); err != nil {
-		logger.Errorf("Unable to setup logger: %s", err)
-	}
+	modeConf.Runner(logConfig)
 
-	if logLevel := os.Getenv(logLevelEnvVar); len(logLevel) > 0 {
-		if err := config.ChangeLogLevel(logLevel); err != nil {
-			logger.Errorf("Unable to change the log level: %s", err)
-		}
-	}
+	metric.AddShutdownMetric(cloudService.GetPrefix(), metricAgent.GetExtraTags(), time.Now(), metricAgent.Demux)
+	lastFlush(logConfig.FlushTimeout, metricAgent, traceAgent, logsAgent)
+}
 
+func setup(mode.Conf) (cloudservice.CloudService, *serverlessInitLog.Config, *trace.ServerlessTraceAgent, *metrics.ServerlessMetricAgent, logsAgent.ServerlessLogsAgent) {
 	tracelog.SetLogger(corelogger{})
 
 	// load proxy settings
-	config.LoadProxyFromEnv(config.Datadog())
+	pkgconfig.LoadProxyFromEnv(pkgconfig.Datadog())
```

Comment:
Do we need to proxy settings in the serverless context?

Reply:
Not sure about that, it wasn't changed from before
https://github.com/DataDog/datadog-agent/blob/main/cmd/serverless-init/main.go#L88

Do you think it is a good idea to merge this PR and address the config refactoring in another?